src/main/java/org/dataone/hashstore/hashfs/HashFileStore.java

```diff
@@ -127,91 +121,124 @@ public HashAddress putObject(InputStream object, String pid, String additionalAl
             if (!checksumAlgorithmSupported) {
-                // TODO: Log failure - include signature values
+                logHashFileStore
+                        .error("HashFileStore.putObject - checksumAlgorithm not supported, checksumAlgorithm: "
+                                + checksumAlgorithm + ". pid: " + pid);
                 throw new IllegalArgumentException(
                         "Checksum algorithm not supported - cannot be used to validate object. checksumAlgorithm: "
                                 + checksumAlgorithm + ". Supported algorithms: "
-                                + Arrays.toString(this.hashUtil.supportedHashAlgorithms));
+                                + Arrays.toString(this.supportedHashAlgorithms));
             }
         }
 
         // Gather HashAddress elements and prepare object permanent address
-        String objAuthorityId = this.hashUtil.getHexDigest(pid, this.objectStoreAlgorithm);
-        String objShardString = this.hashUtil.shard(this.directoryDepth, this.directoryWidth, objAuthorityId);
+        String objAuthorityId = this.getHexDigest(pid, this.objectStoreAlgorithm);
+        String objShardString = this.shard(this.directoryDepth, this.directoryWidth, objAuthorityId);
         String objAbsolutePathString = this.objectStoreDirectory.toString() + "/" + objShardString;
         File objHashAddress = new File(objAbsolutePathString);
         // If file (pid hash) exists, reject request immediately
         if (objHashAddress.exists()) {
+            logHashFileStore.error("HashFileStore.putObject - File already exists for pid: " + pid
+                    + ". Object address: " + objAbsolutePathString);
             throw new FileAlreadyExistsException("File already exists for pid: " + pid);
         }
 
         // Generate tmp file and write to it
-        File tmpFile = this.hashUtil.generateTmpFile("tmp", this.tmpFileDirectory.toFile());
-        Map<String, String> hexDigests = this.hashUtil.writeToTmpFileAndGenerateChecksums(tmpFile, object,
+        logHashFileStore.debug("HashFileStore.putObject - Generating tmpFile");
+        File tmpFile = this.generateTmpFile("tmp", this.tmpFileDirectory);
+        Map<String, String> hexDigests = this.writeToTmpFileAndGenerateChecksums(tmpFile, object,
                 additionalAlgorithm);
 
         // Validate object if checksum and checksum algorithm is passed
         if (checksumAlgorithm != null && checksum != null) {
+            logHashFileStore.info("HashFileStore.putObject - Validating object");
             String digestFromHexDigests = hexDigests.get(checksumAlgorithm);
             if (!checksum.equals(digestFromHexDigests)) {
-                tmpFile.delete();
-                // TODO: Log failure - include signature values
+                // Delete tmp File
+                boolean deleteStatus = tmpFile.delete();
+                if (!deleteStatus) {
```

Comment:
Do we really need to throw an exception when deleting fails? This exception covers the real issue - checksums don't match.

Reply:
if `isNotDuplicate` is false, this means that the move operation failed because the target already exists. My thought process here is that if the file somehow exists when we attempt to move it (even though HashStore logic should have determined that it doesn't before writing to the tmpFile and getting to this point), we should delete the tmpFile (with content written into it) since it will take up unnecessary space on disk. 

If `isNotDuplicate` is true, my intention was to return null values for the HashAddress id, relpath and abspath. It may be useful to have the checksum map in case it's needed so I left that as is. So `storeObject` will always return a HashAddress object - and in Metacat, we would do a get on the `isNotDuplicate` attribute to determine successful storage (where `true` means success). What do you think?